posts/rise-of-microservices-and-docker.md

```diff
@@ -0,0 +1,93 @@
+
+Wait, you might be thinking, what does this even mean?
+
+Now, to understand how we have evolved in the IT infrastructure space, let us take the help of the following diagram:
+
+![Software Infrastructure](../images/microservices-docker-image-four.jpg)
+
+The first picture shows a physical machine or a hardware server. Typically, when we build applications, we use the resources provided by our host OS and the same pattern used to be followed when deploying the application. But what if you want to scale the application? At some point, you might want another hardware server and as the number keeps increasing, so does your cost and other resources like hardware and energy consumptions.
+
+Also, you might be thinking if you really require all the resources of your hardware and host OS at the same time for running your application. Not really, so then why such a massive infrastructure?
+
+This has led to the evolution of hardware virtualization for optimizing IT infrastructure setups through what we call today as Virtual Machines (VMs). As you see in the second diagram, VMs have their guest OS which is run over a single physical machine or host OS. This enables us to run multiple applications without needing installing numerous physical machines. The host OS can ensure that there are systematic resource distribution and load balancing between the different VMs running on it.
+
+Although VMs made software easier to maintain and drastically reduced costs, more optimization was still possible. For instance, not all applications would behave as expected in a guest OS environment. Additionally, the guest OS would require a lot of resources for even running simple processes. 
+
+These problems led to the next innovation: containerization.
+Unlike virtual machines which were more operating system specific, containers are application specific, making them far lighter. Furthermore, VMs can run multiple processes whereas a container runs as a single process. This leads us to two things:
+1. You can run multiple containers on a physical machine, or you can even think of running it on a single VM. In either case, it solves your application related problems
+2. Containerization is not a competition to Virtualization but rather a complementary factor to further optimize your IT software infrastructure
+
+## Docker
+
+Now that we understand the evolution of IT software infrastructure, we might want to know how we can achieve things like `Microservices Architecture` and `Containerization` that we discussed earlier? A simple answer to this could be, `Docker`.
+
+[Docker](https://www.docker.com/) is the world's leading software containerization platform with excellent community support. It allows you to modernize your application and ensures frictionless portability of your applications to cloud infrastructure. It also provides facilities for continuous integration and deployment, managing and scheduling of containers and lot more.
```

Comment:
I think you should include a little more about Docker; this sounds like a sales pitch that I could read off the Docker website ... but it doesn't make it obvious what Docker does and why it will help me.

Reply:
Totally agree with you. So the idea behind this is to get them excited about it and hence the sales pitch. I was planning to write a follow up blog on it. Like discussed earlier, I want to make it a series so the next blog will be completely on Docker and how to use it and also learning to build a dockerized flask application from scratch

Reply:
Agree, I don't think you should leave it on such a cliff-hanger. I would talk a little about some of the benefits of it.